File: wear/src/main/java/com/codegy/aerlink/services/aerlink/reminders/RemindersActivity.java
package com.codegy.aerlink.services.aerlink.reminders;

import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.support.wearable.view.ProgressSpinner;
import android.view.View;
import android.widget.AdapterView;
import android.widget.CheckBox;
import android.widget.ListView;
import android.widget.TextView;
import com.codegy.aerlink.Constants;
import com.codegy.aerlink.R;
import com.codegy.aerlink.utils.AerlinkActivity;

import java.util.List;

public class RemindersActivity extends AerlinkActivity implements ReminderServiceHandler.RemindersCallback, AdapterView.OnItemClickListener {

    private static final String TAG_LOG = RemindersActivity.class.getSimpleName();

    private ReminderCalendar mCalendar;

    private TextView mTitleTextView;
    private ListView mListView;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_reminders);
        setAmbientEnabled();

        Intent intent = getIntent();
        mCalendar = intent.getParcelableExtra(Constants.IE_CALENDAR);

        mDisconnectedLayout = findViewById(R.id.disconnectedLinearLayout);
        mConnectionInfoTextView = (TextView) findViewById(R.id.connectionInfoTextView);

        mConnectionErrorLayout = findViewById(R.id.connectionErrorLinearLayout);

        int calendarColor = Color.parseColor("#"+mCalendar.getColor());

        mLoadingLayout = findViewById(R.id.loadingLayout);
        mLoadingSpinner = (ProgressSpinner) findViewById(R.id.loadingSpinner);
        mLoadingSpinner.setColors(new int[] { calendarColor });

        mTitleTextView = (TextView) findViewById(R.id.titleTextView);
        mTitleTextView.setText(mCalendar.getTitle());
        mTitleTextView.setTextColor(calendarColor);

        mListView = (ListView) findViewById(R.id.listView);
        mListView.setAdapter(new ReminderListAdapter(RemindersActivity.this, null));
        mListView.setOnItemClickListener(RemindersActivity.this);

        mTitleTextView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                ReminderServiceHandler serviceHandler = (ReminderServiceHandler) getServiceHandler(ReminderServiceHandler.class);

                if (serviceHandler != null) {
                    serviceHandler.requestRemindersUpdate(false, new Runnable() {
                        @Override
                        public void run() {
                            setLoading(false);

                            showErrorInterface(true);
                        }
                    });
                }
                else {
                    onDisconnectedFromDevice();
                }
            }
        });

                /*
                SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(RemindersActivity.this);
                String remindersData = sp.getString(Constants.SPK_REMINDER_ITEMS_DATA + mCalendar.getIdentifier(), null);
                if (remindersData != null) {
                    onRemindersUpdated(remindersData);
                }
                */

        updateInterface();
    }

    @Override
    public void onEnterAmbient(Bundle ambientDetails) {
        super.onEnterAmbient(ambientDetails);

        mTitleTextView.setTextColor(Color.LTGRAY);
        ((ReminderListAdapter) mListView.getAdapter()).setAmbient(true);
    }

    @Override
    public void onExitAmbient() {
        super.onExitAmbient();

        mTitleTextView.setTextColor(Color.parseColor("#"+mCalendar.getColor()));
        ((ReminderListAdapter) mListView.getAdapter()).setAmbient(false);
    }

    @Override
    public void onConnectedToDevice() {
        ReminderServiceHandler serviceHandler = (ReminderServiceHandler) getServiceHandler(ReminderServiceHandler.class);

        if (serviceHandler != null) {
            setLoading(true);
            scheduleTimeOutTask();

            serviceHandler.setRemindersCallback(this);
            serviceHandler.requestRemindersUpdate(false, new Runnable() {
                @Override
                public void run() {
                    setLoading(false);

                    showErrorInterface(true);
                }
            });
        }
        else {
            showErrorInterface(true);
        }
    }

    @Override
    public void onDisconnectedFromDevice() {
        super.onDisconnectedFromDevice();

        ReminderServiceHandler serviceHandler = (ReminderServiceHandler) getServiceHandler(ReminderServiceHandler.class);

        if (serviceHandler != null) {
            serviceHandler.setRemindersCallback(null);
        }
    }

    @Override
    public void onDataTransferStarted() {
        cancelTimeOutTask();
    }

    @Override
    public void onRemindersUpdated(final List<ReminderItem> reminders) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                setLoading(false);
                showErrorInterface(false);
                cancelTimeOutTask();

                if (mListView != null) {
                    ((ReminderListAdapter) mListView.getAdapter()).refresh(reminders);
                }
            }
        });
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        final ReminderServiceHandler serviceHandler = (ReminderServiceHandler) getServiceHandler(ReminderServiceHandler.class);

        if (serviceHandler != null) {
            CheckBox checkBox = (CheckBox) view.findViewById(R.id.checkBox);
            if (checkBox != null) {
                final boolean completed = !checkBox.isChecked();
                checkBox.setChecked(completed);

                setLoading(true);

                final ReminderItem item = ((ReminderListAdapter) mListView.getAdapter()).getItem(position);
                item.setCompleted(completed);
                serviceHandler.setReminderCompleted(item,
                        new Runnable() { // success
                            @Override
                            public void run() {
                                setLoading(false);
                                showErrorInterface(false);

                                serviceHandler.updateCachedData(mCalendar.getIdentifier(), item);
                            }
                        },
                        new Runnable() { // failure
                            @Override
                            public void run() {
                                setLoading(false);

                                showErrorInterface(true);
                            }
                        }
                );
            }
        }
        else {
            onDisconnectedFromDevice();
        }
    }

}

File: wear/src/main/java/com/codegy/aerlink/ui/reminders/RemindersActivity.kt
package com.codegy.aerlink.ui.reminders

import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.codegy.aerlink.R
import com.codegy.aerlink.service.aerlink.reminders.RemindersServiceManager
import com.codegy.aerlink.service.aerlink.reminders.model.ReminderCalendar
import com.codegy.aerlink.service.aerlink.reminders.model.ReminderItem
import com.codegy.aerlink.ui.AerlinkActivity
import kotlinx.android.synthetic.main.activity_reminders.*
import kotlinx.android.synthetic.main.list_item_reminder.view.*

class RemindersActivity : AerlinkActivity(), RemindersServiceManager.RemindersCallback {
    private val calendar: ReminderCalendar?
        get() = intent?.getParcelableExtra(IE_CALENDAR)

    inner class Adapter: RecyclerView.Adapter<Adapter.ViewHolder>() {
        private var reminders: List<ReminderItem> = listOf()

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            val layoutInflater: LayoutInflater = LayoutInflater.from(parent.context)
            val itemView = layoutInflater.inflate(R.layout.list_item_reminder, parent, false)
            return ViewHolder(itemView)
        }

        override fun getItemCount(): Int = reminders.size

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            val reminder = reminders[position]
            holder.checkBox.text = reminder.title
            holder.checkBox.isChecked = reminder.isCompleted
            if (isAmbient) {
                holder.checkBox.setTextColor(Color.LTGRAY)
            } else {
                holder.checkBox.setTextColor(Color.WHITE)
            }
            holder.itemView.setOnClickListener {
                val newValue = !holder.checkBox.isChecked
                val calendarIdentifier = calendar?.identifier ?: return@setOnClickListener
                val serviceManager = getServiceManager(RemindersServiceManager::class) ?: return@setOnClickListener
                serviceManager.setReminderCompleted(newValue, reminder, calendarIdentifier) {
                    holder.checkBox.post {
                        holder.checkBox.isChecked = reminder.isCompleted
                    }
                }
            }
        }

        fun reloadData(reminders: List<ReminderItem>) {
            this.reminders = reminders
            notifyDataSetChanged()
        }

        inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
            val checkBox: CheckBox = itemView.checkBox
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_reminders)
        setAmbientEnabled()
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = Adapter()
        updateUI()
    }

    override fun onPause() {
        val serviceManager = getServiceManager(RemindersServiceManager::class)
        serviceManager?.remindersCallback = null
        super.onPause()
    }

    override fun onNewIntent(intent: Intent?) {
        super.onNewIntent(intent)
        setIntent(intent)
        updateUI()
        updateReminders()
    }

    private fun updateUI() {
        val calendar = calendar ?: return
        val calendarColor = Color.parseColor("#" + calendar.color)
        titleTextView.text = calendar.title
        titleTextView.setTextColor(calendarColor)
    }

    private fun updateReminders() {
        val serviceManager = getServiceManager(RemindersServiceManager::class) ?: return
        calendar?.let {
            showLoading(true)
            serviceManager.requestRemindersUpdate(false, it.identifier)
        } ?: run { finish() }
    }

    override fun onEnterAmbient(ambientDetails: Bundle?) {
        super.onEnterAmbient(ambientDetails)
        recyclerView.adapter?.notifyDataSetChanged()
    }

    override fun onExitAmbient() {
        super.onExitAmbient()
        recyclerView.adapter?.notifyDataSetChanged()
    }

    override fun onConnectedToDevice() {
        val serviceManager = getServiceManager(RemindersServiceManager::class)
        if (serviceManager == null) {
            Log.v(LOG_TAG, "Service manager not available")
            showError(true)
            return
        }

        serviceManager.remindersCallback = this
        updateReminders()
    }

    override fun onError() {
        showLoading(false)
        showError(true)
    }

    override fun onDataTransferStarted() {
        showLoading(true)
    }

    override fun onRemindersUpdated(reminders: List<ReminderItem>) {
        runOnUiThread {
            showLoading(false)
            val adapter = recyclerView.adapter as? Adapter
            adapter?.reloadData(reminders)
        }
    }

    companion object {
        private val LOG_TAG = RemindersActivity::class.java.simpleName
        const val IE_CALENDAR: String = "com.codegy.aerlink.ui.reminders.IE_CALENDAR"

        fun startWithCalendar(context: Context, calendar: ReminderCalendar) {
            val intent = Intent(context, RemindersActivity::class.java)
            intent.putExtra(IE_CALENDAR, calendar)
            context.startActivity(intent)
        }
    }
}
